From 4396b99cfc1efae431e47d51abf183896ece5c3b Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:05:30 -0400
Subject: [ethernet] qlcnic: Fix memory allocation

Message-id: <1408626422-3941-47-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88642
O-Subject: [PATCH RHEL 7.1 046/138] qlcnic: Fix memory allocation
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Manish Chopra <manish.chopra@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id 3fc38e267bfbea5e33e2222d6babc3b06c2bb642

o Use vzalloc() instead of kzalloc() for allocation of
  bootloader size memory. kzalloc() may fail to allocate
  the size of bootloader

Signed-off-by: Manish Chopra <manish.chopra@qlogic.com>
Signed-off-by: Himanshu Madhani <himanshu.madhani@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_init.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_init.c
index fae1b71..cac0503 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_init.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_init.c
@@ -1254,24 +1254,24 @@ static int qlcnic_83xx_copy_bootloader(struct qlcnic_adapter *adapter)
  if (size & 0xF)
   size = (size + 16) & ~0xF;
 
- p_cache = kzalloc(size, GFP_KERNEL);
+ p_cache = vzalloc(size);
  if (p_cache == NULL)
   return -ENOMEM;
 
  ret = qlcnic_83xx_lockless_flash_read32(adapter, src, p_cache,
       size / sizeof(u32));
  if (ret) {
-  kfree(p_cache);
+  vfree(p_cache);
   return ret;
  }
  /* 16 byte write to MS memory */
  ret = qlcnic_83xx_ms_mem_write128(adapter, dest, (u32 *)p_cache,
        size / 16);
  if (ret) {
-  kfree(p_cache);
+  vfree(p_cache);
   return ret;
  }
- kfree(p_cache);
+ vfree(p_cache);
 
  return ret;
 }
-- 
1.7.1